semantic/poly_types.go

```diff
@@ -23,12 +23,16 @@ type TypeExpression interface {
 // PolyType represents a polymorphic type, meaning that the type may have multiple free type variables.
 type PolyType interface {
 	TypeExpression
+
 	// occurs reports whether tv is a free variable in the type.
 	occurs(tv Tvar) bool
-	// substituteType replaces tv for t producing a new type.
-	substituteType(tv Tvar, t PolyType) PolyType
+
+	// apply applies a substitution to a type
+	// apply returns false if there were no substitutions made
+	apply(sub Substitution) (PolyType, bool)
+
 	// unifyType unifies the two types given the kind constraints and produces a substitution.
```

Comment:
This comment should be updated to reflect that it will return the possibly unchanged, or possible expanded set of substitutions.

Reply:
Resolved.